File: Chess/Bitboards.cs
using System.Dynamic;
using System.Numerics;
using System.Reflection.Metadata;

namespace Chess {
    public static class Bitboards {
        
        public const ulong FirstRankMask   = 0x00000000000000FF;
        public const ulong SecondRankMask  = 0x000000000000FF00;
        public const ulong ThirdRankMask   = 0x0000000000FF0000;
        public const ulong FourthRankMask  = 0x00000000FF000000;
        public const ulong FifthRankMask   = 0x000000FF00000000;
        public const ulong SixthRankMask   = 0x0000FF0000000000;
        public const ulong SeventhRankMask = 0x00FF000000000000;
        public const ulong EighthRankMask  = 0xFF00000000000000;

        public const ulong NotFirstFileMask   = 0xFEFEFEFEFEFEFEFE;
        public const ulong NotEighthFileMask   = 0x7F7F7F7F7F7F7F7F;

        public const ulong Board6x6 = 0x7E7E7E7E7E7E00;

        public const ulong WhiteKingSideCastlingMask = 0x60;
        public const ulong BlackKingSideCastlingMask = 0x6000000000000000;
        public const ulong WhiteQueenSideCastlingMask = 0xE;
        public const ulong BlackQueenSideCastlingMask = 0xE00000000000000;

        public const ulong FullBoardMask   = 0xFFFFFFFFFFFFFFFF;


        public static bool IsSquareOccupied(ulong bitboard, int square) {
            return ((bitboard >> square) & 1) == 1;
        }

        public static int GetLS1BSquare(ulong bitboard) {
            return BitOperations.TrailingZeroCount(bitboard);
        }

        public static ulong GetLS1BBit(ulong bitboard) {
            return 1UL << GetLS1BSquare(bitboard);
        }

        public static int SquareFromBit(ulong bit) {
            return GetLS1BSquare(bit);
        }

        public static ulong BitFromSquare(int square) {
            return 1UL << square;
        }

        public static ulong Shift(ulong bitboard, int squares) {
            if (squares > 0) {
                bitboard <<= squares;
            } else {
                bitboard >>= -squares;
            }
            return bitboard;
        }

        public static ulong GetNorthOffset(ulong bitboard, int numSquares) {
            return bitboard << (8 * numSquares);
        }

        public static ulong GetSouthOffset(ulong bitboard, int numSquares) {
            return bitboard >> (8 * numSquares);
        }

        public static ulong GetWestOffest(ulong bitboard, int numSquares) {
            return bitboard << (1 * numSquares);
        }

        public static ulong GetEastOffset(ulong bitboard, int numSquares) {
            return bitboard >> (1 * numSquares);
        }

        public static ulong GetNorthWestOffset(ulong bitboard, int numSquares) {
            return bitboard << (7 * numSquares);
        }

        public static ulong GetNorthEastOffset(ulong bitboard, int numSquares) {
            return bitboard << (9 * numSquares);
        }

        public static ulong GetSouthWestOffset(ulong bitboard, int numSquares) {
            return bitboard >> (9 * numSquares);
        }

        public static ulong GetSouthEastOffset(ulong bitboard, int numSquares) {
            return bitboard >> (7 * numSquares);
        }

        public static ulong GetOffset(ulong bitboard, int numSquares, int direction) {
            return direction switch
            {
                8 => GetNorthOffset(bitboard, numSquares),
                -8 => GetSouthOffset(bitboard, numSquares),
                -1 => GetWestOffest(bitboard, numSquares),
                1 => GetEastOffset(bitboard, numSquares),
                7 => GetNorthWestOffset(bitboard, numSquares),
                9 => GetNorthEastOffset(bitboard, numSquares),
                -9 => GetSouthEastOffset(bitboard, numSquares),
                -7 => GetSouthWestOffset(bitboard, numSquares),
                _ => 1,
            };
        }

        public static int GetBitsCount(ulong bitboard) {
            int count = 0;
            while (bitboard != 0) {
                count++;
                bitboard &= ~GetLS1BBit(bitboard);
            }
            return count;
        }
    }
}

File: chess/src/bitboards.rs
// binary masks
pub const FIRST_RANK_MASK:   u64 = 0x00000000000000FF;
pub const SECOND_RANK_MASK:  u64 = 0x000000000000FF00;
pub const FOURTH_RANK_MASK:  u64 = 0x00000000FF000000;
pub const FIFTH_RANK_MASK:   u64 = 0x000000FF00000000;
pub const SEVENTH_RANK_MASK: u64 = 0x00FF000000000000;
pub const EIGHTH_RANK_MASK:  u64 = 0xFF00000000000000;

pub const NOT_FIRST_FILE_MASK: u64 = !0x8080808080808080;
pub const NOT_EIGHTH_FILE_MASK: u64 = !0x101010101010101;

/// The entire board except all edges (ranks 1, 8 and files a, h)
pub const BOARD_6X6_MASK: u64 = 0x7E7E7E7E7E7E00;

// castling masks
// contain the squares that are needed to be checked in order to make castling a legal move
pub const WHITE_KING_SIDE_CASTLING_MASK:            u64 = 0x6;
pub const BLACK_KING_SIDE_CASTLING_MASK:            u64 = 0x600000000000000;
pub const WHITE_QUEEN_SIDE_CASTLING_MASK:           u64 = 0x70;
pub const BLACK_QUEEN_SIDE_CASTLING_MASK:           u64 = 0x7000000000000000;
pub const WHITE_QUEEN_SIDE_CASTLING_KING_PATH_MASK: u64 = 0x30;
pub const BLACK_QUEEN_SIDE_CASTLING_KING_PATH_MASK: u64 = 0x3000000000000000;

pub fn get_ls1b(bitboard: u64) -> usize {
    bitboard.trailing_zeros() as usize
}

/// Returns a bitboard with a single set bit from the provided square
pub fn get_bit_from_square(square: usize) -> u64 {
    1u64 << square
}

/// Returns the provided bitboard shifted by the provided square number upwards
pub fn shift_bb(bitboard: u64, shift_amount: i8) -> u64 {
    debug_assert!(shift_amount >= -63 && shift_amount <= 63, "Shift amount out of bounds");
    if shift_amount > 0 {
        bitboard << shift_amount
    } else {
        bitboard >> -shift_amount
    }
}